src/lock_free/mwcas_descriptor.cpp

```diff
@@ -182,49 +191,35 @@ MwCASDescriptor::MwCASInternal(  //
   }
 
   const auto succeeded = (cur_stat == kSucceeded);
-  auto follow_cnt = 0U;
   if (succeeded) {
     for (size_t i = 0; i < target_cnt_; ++i) {
       auto &target = targets_[i];
       const auto ver = (target.old_val.load(kRelaxed) + kVersionUnit) & kVersionMask;
-      follow_cnt += Finalize(base_addr, target, (target.new_val | ver));
+      Finalize(base_addr, target, (target.new_val | ver));
     }
   } else {
     for (size_t i = 0; i < target_cnt_; ++i) {
       auto &target = targets_[i];
       const auto val = (target.old_val.load(kRelaxed) + kVersionUnit) & kVerAndValMask;
-      follow_cnt += Finalize(base_addr, target, val);
+      Finalize(base_addr, target, val);
     }
   }
 
-  if (follow_cnt == 0 || exit_cnt_.fetch_add(1, kRelaxed) == follow_cnt) {
-    _tls.reset(this);
-  }
-
   return succeeded;
 }
 
-auto
+void
 MwCASDescriptor::Finalize(  //
     const uint64_t desc_addr,
     MwCASTarget &target,
     const uint64_t desired)  //
-    -> uint32_t
 {
   while (true) {
     auto expected = target.addr->load(kRelaxed);
-    auto cur_cnt = target.cnt.load(kRelaxed);
-    const auto cnt = static_cast<uint32_t>((expected & kCntMask) >> kCntShift);
-    if (((expected ^ desc_addr) & kDescMask)                            // already swapped, or
-        || (cur_cnt == cnt                                              // counter is latest and
-            && target.addr->compare_exchange_strong(expected, desired,  // swap succeeds
-                                                    kRelaxed, kRelaxed))) {
-      return cur_cnt;
-    }
-    // the descriptor's counter should be updated
-    while (cur_cnt < cnt) {
-      if (target.cnt.compare_exchange_weak(cur_cnt, cnt, kRelaxed, kRelaxed)) break;
-      CPP_UTILITY_SPINLOCK_HINT
+    if (((expected ^ desc_addr) & kDescMask)                        // already swapped, or
+        || target.addr->compare_exchange_strong(expected, desired,  // swap succeeds
+                                                kRelaxed, kRelaxed)) {
+      return;
     }
   }
```

Comment:
カウンタ関係の処理がなくなり即時リトライできるようになったので、 `load` は最初の1回のみでOKです（CAS失敗時に `expected` の値が最新値に更新されるため）。また、リトライ前提のときは `_weak` の方のCASを使います（より効率的なCPU命令が使われる **場合がある** [^1]）。

```suggestion
  auto expected = target.addr->load(kRelaxed);
  while (((expected ^ desc_addr) & kDescMask) == 0
      && !target.addr->compare_exchange_weak(expected, desired, kRelaxed, kRelaxed)) {
    CPP_UTILITY_SPINLOCK_HINT
  }
```

[^1]: 現実的（IntelやAMDのCPUを使う分）には同じCPU命令が使われると思いますが。